### PR TITLE
Fixes literally not forcing re-dispatch for inline='always'

### LINKED ISSUE #5887: Regression with literally and overloads with inline='always'
- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Hi, 

While trying to migrate to Numba 0.50 we encountered the following problem with literally that may now fail if called from overloads compiled with inline='always'. The error itself looks like this:
```
DEBUG: dtype= unicode_type
Traceback (most recent call last):
  File "test_inline_literally.py", line 36, in <module>
    res = test_impl(data, dtype)
  File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\dispatcher.py", line 401, in _compile_for_args
    error_rewrite(e, 'typing')
  File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\dispatcher.py", line 344, in error_rewrite
    reraise(type(e), e, None)
  File "C:\Users\akozlov\AppData\Local\Continuum\anaconda3\numba_master\numba\numba\core\utils.py", line 79, in reraise
    raise value.with_traceback(tb)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at <numba.core.typeinfer.CallConstraint object at 0x0000023CF5AA24F0>.
'NoneType' object is not callable
[1] During: resolving callee type: Function(<function foo at 0x0000023CF24E5280>)
[2] During: typing of call at test_inline_literally.py (32)

Enable logging at debug level for details.

File "test_inline_literally.py", line 32:
def test_impl(arr, dtype):
    return foo(arr, dtype)
    ^
```

And below is the reproducer:
```
import numba
import numpy as np
from numba.extending import overload
from numba import types
from numba import literally


def foo(val, dtype):
    pass

@overload(foo, inline='always')    # seen with inline='always' only
def foo_ovld(arr, dtype):

    print("DEBUG: dtype=", dtype)
    if not isinstance(dtype, types.StringLiteral):
        def foo_noop(arr, dtype):
            return literally(dtype)
        return foo_noop

    if dtype.literal_value == 'str':
        def foo_as_str_impl(arr, dtype):
            return [str(x) for x in arr]
        return foo_as_str_impl

    if dtype.literal_value in ('int64', 'float64'):
        def foo_as_num_impl(arr, dtype):
            return list(arr.astype(np.dtype(dtype)))
        return foo_as_num_impl

@numba.njit
def test_impl(arr, dtype):
    return foo(arr, dtype)

data = np.asarray([1, 2, 3, 2, 4], dtype=np.int32)
dtype = 'int64'
print(test_impl(data, dtype))
```

The issue was found on 0.50.0rc1 but still remains in 0.51.0.dev0+56.g92df8df57
It seems it is somehow related to #5673 as it's not observed before it, but please note that even though test from our code base always passes on 0.49.1 the above script which is just simplified version might occasionally fail with 'Repeated literal typing request' as below:
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_inline_literally_2.py", line 20, in <module>
    res = test_impl(data, dtype)
  File "c:\users\akozlov\appdata\local\continuum\anaconda3\numba_master\numba\numba\core\dispatcher.py", line 376, in _compile_for_args
    return self._compile_for_args(*args)
  File "c:\users\akozlov\appdata\local\continuum\anaconda3\numba_master\numba\numba\core\dispatcher.py", line 369, in _compile_for_args
    raise errors.CompilerError(m.format(info))
numba.core.errors.CompilerError: Repeated literal typing request.
Arg #1 is Literal[str](int64).
This is likely caused by an error in typing. Please see nested and suppressed exceptions.
```
Not sure if this latter is a known issue and wasn't fixed already (please confirm if it's so).
But the main issue is of course new one _'NoneType' object is not callable_ problem.

Best Regards,
Alexey.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #5887

### CODE REVIEW

File: numba/core/typing/templates.py [hunk 1]
```diff
@@ -517,7 +517,20 @@ def generic(self, args, kws):
...
+            # happen and continue to `resolve` as that will trigger a
+            # ForceLiteralArg exception instance that is appropriately formed
+            # for consumption in the dispatcher.
+            try:
+                ir = inline_worker.run_untyped_passes(
+                    disp_type.dispatcher.py_func)
+            except errors.ForceLiteralArg:
+                pass
             resolve = disp_type.dispatcher.get_call_template
             template, pysig, folded_args, kws = resolve(new_args, kws)
```
Comment (sklam) [inline numba/core/typing/templates.py:535]:
```suggestion
            except errors.ForceLiteralArg:
                resolve = disp_type.dispatcher.get_call_template
                template, pysig, folded_args, kws = resolve(new_args, kws)
                raise errors.CompilerError("unreachable")
                 
            resolve = disp_type.dispatcher.get_call_template
            template, pysig, folded_args, kws = resolve(new_args, kws)
```

Comment (stuartarchibald) [reply numba/core/typing/templates.py:535]: following more out of band discussion, going to try swapping the order of the resolve and the pipieline call

Comment (stuartarchibald) [reply numba/core/typing/templates.py]: done in https://github.com/numba/numba/pull/5889/commits/c0e231b3c93136432067125b8cc600739c5ae692

**Comment by stuartarchibald** [comment]
I approve 34294d6, thanks.

**Pull request merged**